lib/src/v3/variable_tokenizer.dart

```diff
@@ -0,0 +1,292 @@
+import 'package:charcode/charcode.dart';
+
+import 'query_description.dart';
+import 'types.dart';
+
+/// In addition to indexed variables supported by the postgres protocol, most
+/// postgres clients (including this package) support a variable mode in which
+/// variables are identified by `@` followed by a name and an optional type.
+///
+/// Since this format is not natively understood by postgres, a translation
+/// layer needs to happen in Dart.
+/// A flexible [_variableCodeUnit] can be set to start a variable (we'll
+/// assume `@`, the default, here). Then, a variable can be used with `@varname`.
+/// Variable names can consist of alphanumeric characters and underscores.
+/// As part of a variable, the type can also be declared (e.g. `@x:int8`). Valid
+/// type names are given by [PgDataType.nameForSubstitution].
+///
+/// Just like postgres, we ignore variables inside string literals, identifiers
+/// or comments.
+class VariableTokenizer {
```

Comment:
On the surface, this is good, and I don't have much issue with it. However, I fear that without a proper grammar and parser, it is likely that this will have an edge case that will get the escape wrong. Like a commented-out string literal without a closing quote...

I don't think it is possible to handle all use cases here (within reasonable complexity), and it may be outside of the scope of the library to maintain a grammar and parser for it. But it may be good to put a note about it somewhere.

Reply:
I've added a lot of documentation to the `PgSql` class and its constructors. `PgSql.map` points out that the transformation may not be correct in edge cases like invalid statements.